src/odoc/resolver.ml

```diff
@@ -410,12 +439,60 @@ let add_unit_to_cache u =
   in
   Hashtbl.add unit_cache target_name [ u ]
 
+let lookup_path _ap ~pages ~libs:_ ~hierarchy (kind, tag, path) =
+  let module Env = Odoc_xref2.Env in
+  let ( >>= ) x f = match x with Some x' -> f x' | None -> None in
+  let page_path_to_path path =
+    (* Turn [foo/bar] into [foo/page-bar.odoc]. *)
+    let segs =
+      match List.rev path with
+      | [] -> []
+      | name :: rest -> List.rev (("page-" ^ name ^ ".odoc") :: rest)
+    in
+    Fs.File.of_segs segs
+  in
+  let find_by_path ?root named_roots path =
+    match Named_roots.find_by_path ?root named_roots ~path with
+    | Ok x -> x
+    | Error (NoPackage | NoRoot) -> None
+  in
+  let load_page path =
+    match load_unit_from_file path with
+    | Some (Odoc_file.Page_content page) -> Env.Path_page page
+    | _ -> Env.Path_not_found
+  in
+  let find_page ?root path =
+    (pages >>= fun pages -> find_by_path ?root pages path) |> function
+    | Some path -> load_page path
+    | None -> Env.Path_not_found
+  in
+  let find_page_in_hierarchy path =
+    match hierarchy with
+    | Some hierarchy -> (
+        match Hierarchy.resolve_relative hierarchy path with
+        | Ok path -> load_page path
+        | Error `Escape_hierarchy ->
+            Env.Path_not_found (* TODO: propagate more information *))
```

Comment:
Writing a comment here not to forget this TODO.

Reply:
This will be part of improving error propagation in the resolver (cf comment https://github.com/ocaml/odoc/pull/1150#discussion_r1673755882) that I would prefer to think about later.